1 denote logseq migration

hapst3r opened the pull request:
You suggested I make a pull request after I am satisfied with how it is working.

I deleted the superfluous code (denote silos to logseq), because I think this will be solved either be someone else or by us at a later point in time (maybe with another branch, let's see).

hapst3r committed:
nm/denote-to-logseq: init

hapst3r committed:
nm/denote-to-logseq: Wrote RX matching denote-link

I am still trying to wrap my head around opening and writing buffers.  For starters,
I wrote a RX expression that I believe to be able to 100% identify a denote link.

This is important because the DENOTE and LOGSEQ link format seems to be quite
different. DENOTE has [[denote:ID][DESC]], LOGSEQ has [[FILE TITLE]] without
description.  Thus I think it is important to be specific with RX.

Next up is the identification of the file DENOTE links TO, getting this file's TITLE.

hapst3r committed:
nm/denote-to-logseq: Clean slate, collecting elements

I realized that I lose focus when I always see the functions initially in place.  In
contrast, I will try to work my way up from nothing and peek at your approach as necessary.

hapst3r committed:
nm/denote-to-logseq: (temporarily) add outline delimiters

;;; et cetera help keep the document structured.  Will use that so I don't get lost

hapst3r committed:
nm/denote-to-logseq: Tested REPLACE-REGEXP fn

If I am not mistaken, this is how I am going to replace the initial denote link with
a logseq link.

What I will have to find out next is how I am going to get according file and its
TITLE.

hapst3r committed:
nm/denote-to-logseq: Identifying and prototyping steps towards my goal

I walk them baby steps to not get lost, maybe this will pay off :)

hapst3r committed:
nm/denote-to-logseq: Identifying the next step

Now I just have to find out how to use ORG-ELEMENT :D

hapst3r committed:
nm/denote-to-logseq: Wild experimentation

hapst3r committed:
nm/denote-to-logseq: First attempt at writing the link-conversion fn

hapst3r committed:
nm/denote-to-logseq: Getting closer I guess?

So I have written NM--CONVERT-DENOTE-LINKS-TO-LOGSEQ that does most of the things I
want corrctly, whereby I mean that all values in the LET-expressions are correct.
However, there is an issue with matching and replacing that I don't quite understand.

Currently, I have one magic variable baked in, MY-DENOTE-DIRECTORY.  I will have to
find a way how to eliminate it, though I think this is less important than the
matching/replacing part.

Also, the file is a total mess overall, will have to clean up soon™.

hapst3r committed:
nm/denote-to-logseq: Approaching a working prototype

So I found out that I can use denote's internal machinery to great effect, which I
did.  Now, I believe to have all the pieces I need in order to replace a denote link
by a logseq link.

However, for a reason that I don't understand, the combination of `re-search-forward`
and `replace-match` does not work as I would expect it to work.

I will have to investigate this further tomorrow.

hapst3r committed:
nm/denote-to-logseq: Trying to resolve the match/replace issue

So I thought I might find an ORG-function with which I can access the link under
point and just change the link contents, without resorting to regex replacement.
Unfortunately, I was unable to find that.

I honestly don't understand the error that's thrown (args out of range) and how to
mitigate it.  Strangely enough, I am able to successfully replace ONE link when I
insert a `re-search-backward` before the `replace-match`.  However, when I wrap the
whole shebang in a `while` instead of an `if`, the execution does not terminate.

hapst3r committed:
nm/denote-to-logseq: Reaching capacity and motivation limits

Hej.  After even more time spent unsuccessfully trying to get things working, I think
I have reached my limits.  It is just that I don't see how to achieve what I want to
achieve.

These are the ventures I tried but which proved fruitless:
(1) I thought there might be a non-interactive facility to act on an org-link.  That
would have been the easiest, but I couldn't find anything like this.
(2) I have tried different constructs for both searching and replacing data.  The
commit before this one shows an approach which appears conceptually sound but doesn't
work: it yields an arg-out-of-range error which I have not been able to resolve.
(2.1) I tried to wrap the let block in a `save-match-data` block because I suspected
the reason to be that after some transformations in `let`, match data would've been
lost. Didn't work.
(2.2) I tried to use `replace-string` like you did in your function, but this doesn't
seem to work with regular expressions.

As I said, I think I will have to slow down working on this, because I have sat on
this problems for an eternity, facing problems I currently cannot surmount.  I might
contact you soon via matrix and/or ask around the emacs matrix channel.

If you have any ideas regarding my problems especially in the last two commits, I'd
be grateful.

bitspook committed:
Merge remote-tracking branch 'origin/master' into 1-denote-logseq-migration

bitspook committed:
feat: basic denote to logseq migration

Instead of solving the regexp-replace problem (which I am sure would be
interesting and difficult), I chose to take a slightly different approach of
deleting the existing denote org-link and inserting a logseq-link in its place.

Also took some liberties to move some code around and rename some things, sorry
@hapst3r

A call to

```lisp
(let ((logseq-dir "/tmp/logseq"))
  (nm--migrate-denote-to-logseq))
```

works for all my notes, so I assume it is good enough for initial iteration.

hapst3r committed:
nm/denote-to-logseq: fix: Proper new link inserting and and directory substitution

Testing your new code, no files would be written to a new directory.  However, the
links would be changed in the old directory (thank god I have used "test-*"
directories!), albeit without inserting a space after.

I thus checked the functioning of `nm--migrate-denote-files-to-logseq` and found that
`expand-file-name` does not behave as expected in combination with your rendition of
`string-replace`.  Using LET to input values for `denote-directory` and
`logseq-directory`, I found that DEST evaluates to ./path-to-file.org.

I then remembered that there is a native function that returns just the file name
without directory, `file-name-nondirectory`.  Using these changes, I was able to
successfully migrate from TEST-DENOTE to TEST-LOGSEQ :)

I don't know how exactly this works, but I know that it has not worked before and
works afterwards with the same input values.

Going forward, what do you think is required to improve on the version that is
already present?  One thing I have in mind but with which I don't have any experience
whatsoever is emitting error messages when stuff doesn't work as expected.  Then, I
feel like returning a list containing N NIL where N is the number of files is kind of
noisy.

Furthermore, I believe the next step would be to check if there is a way to refactor
so that there is less redundance for the two migration functions (roam -> denote,
denote -> logseq).  Again, that is an idea, but I am not experienced with this kind
of stuff.

Two closing thoughts: first off, at what point do you think it would make sense to
post and update to reddit (could increase exposure and more skilled people joining
in)? Second, what would be required to make the program, as is, usable as a package?
I feel like the minimum requirement for that would be the capacity to define
LOGSEQ-DIRECTORY and DENOTE-DIRECTORY and to emit an error message when they are not
defined or no files are found there.

hapst3r committed:
nm/denote-to-logseq: Cleaning up

hapst3r committed:
nm/denote-to-logseq: Cleaner code

I removed CONTENTS in LET* because it wasn't used and served no purpose.

I implemented the solution proposed by yantar92, which meant to replace
`delete-region` and `insert` with `replace-region-contents`.

Finally, I thought it a good idea to also steal (from yantar92) the
`org-fill-paragraph` function at the end so that the file doesn't look ugly after
replacing a long (denote) link with a much shorter (logseq) link.  Dunno if that
matters though.

hapst3r committed:
nm/denote-to-logseq|feature: Migrate silos pt. 1

Since I solely work with denote silos -- denote subdirs that don't share keywords
with one another --, I'd like to produce a function that is capable of migrating
silos.  I don't have the logic completely planned out, but I will work on that :)

hapst3r committed:
nm/denote-to-logseq: fix: Whitespace between link and text

hapst3r committed:
nm/denote-to-logseq: Remove specialist code related to silos

You are right that my initial use case was too personal to be of general use.  So I
will delete this fragment in favour of a future attempt as described in my previous
message.

bitspook commented on denote-to-logseq.el:
```diff
@@ -0,0 +1,63 @@
+;; Package-Requires: ((emacs "27.2") (denote "1.0.0")
+;;
+;;; Commentary:
+;; A WIP package for migrating Denote notes to Logseq
```
Is this still a work-in-progress or are you ready to merge it into master (main) branch? If you're ready, please change this comment to remove the WIP mention.

hapst3r replied:
I'll be honest, I'd prefer if you decided this. I honestly don't know which criteria it would have to fulfill to be ready.

However, given that migration now respects denote silos (see last two commits and below), it features everything I'd want it to have - at least for the moment.

bitspook replied:
> it features everything I'd want it to have - at least for the moment.

That is the criteria that matters imo. 

That's how I usually approach the open-source software I build. Write for yourself, present it well enough that someone with *exact same* use-case can use it too. If someone wants more from the software, they can communicate (e.g create a Github issue); and then you can decide if you have time and energy to extend it and in which manner. Otherwise they can always fork it and make their own changes.

bitspook commented on denote-to-logseq.el:
```diff
@@ -0,0 +1,63 @@
+Copy the file from `denote-directory' to `logseq-directory' and
+convert all denote links to logseq."
+  (let ((dest (expand-file-name
+	       (file-name-nondirectory file)
```
I suppose this call here is the cause of notes getting flattened when migrated from `denote-directory` to `logseq-directory`. I think we should remedy this before merging. What do you say?

hapst3r replied:
Yeah, you are totally right. I hope to have found a working solution, see the last two commits.

bitspook commented on denote-to-logseq.el:
```diff
@@ -0,0 +1,63 @@
+      (nm--convert-denote-links-to-logseq)
+      (write-file dest))))
+
+(defun nm--ensure-dir-path (path)
```
This function served its purpose when L:40 created destination path of a migrated note by replacing `denote-directory` in note's filepath with `logseq-directory`. Since we are not doing that anymore, I think this function is causing more possible problems while solving none (paths can be tricky to work with across platforms).

In current state of code, I think we should remove this function altogether.

hapst3r replied:
I honestly have not been sure what its original purpose has been, so if you think we can get rid of it, I'm all for it :)

bitspook commented on denote-to-logseq.el:
```diff
@@ -0,0 +1,63 @@
+`logseq-directory'."
+  (let* ((denote-directory (nm--ensure-dir-path denote-directory))
+         (logseq-directory (nm--ensure-dir-path logseq-directory))
+         (denote-org-files (directory-files denote-directory t "\.org$")))
```
Do logseq support non-org files as well? Do people have things like images and other files in the `denote-directory`? If yes, I think we should handle those as well by simply copying them over to `logseq-directory`.

hapst3r replied:
Well, I honestly don't know, since it is not my use-case. And since I don't know, I think I would prefer to postpone this feature and wait for people to request and/or implement it.

What do you think?

bitspook replied:
>  I think I would prefer to postpone this feature and wait for people to request and/or implement it.

Seems reasonable.

bitspook reviewed (changes requested):
Hey!

Thank you for writing all this code! I know you want to release the kraken asap, but I've left some comments where I thought some changes might be needed. Please take a look.

hapst3r committed:
fix: Starting a logic to detect a non-DENOTE-DIRECTORY-subdir

I am starting to figure out a logic that should detect whether we are dealing with a
silo or not.

hapst3r committed:
feature: Detect a silo'd file and act accordingly (not finished)

So I wrote functions that detect if file is not in DENOTE-DIRECTORY root but in a
subdirectory, and act accordingly: produce a fitting LOGSEQ-SILO (LOGSEQ-DIRECTORY +
silo/subdir) string, mkdir (if not existing), write file to DEST.

After a short moment of triumph I realized that even though the logic of this
function is sound, it does not work well with the logic in place for
NM--MIGRATE-DENOTE-TO-LOGSEQ.  This function relies on the fact that the
DENOTE-DIRECTORY contains org files, which it doesn't, both in my use case and the
present test case.

This cannot be too complicated, but I don't have the resources to finish it today.

hapst3r committed:
feature: Consider denote silos when migrating

I knew it couldn't be that hard.  I found that `directory-files-recurssively` does
exactly what I want, a long list with all files in the denote directory (including
subdirs).

I think this kind of concludes the addition of this feature for now.

bitspook reviewed (approved):
Changes look good to me. I am going to merge this branch, and then I'll some documentation to the readme to document how to use this script. 

Thanks @hapst3r !
